### DISCUSSION TITLE
Allow per scenario differing ValueRetrievers in combination with the TableHelperExtensionMethods methods in parallel executed tests.

### DISCUSSION THREAD

**Discussion by ajeckmans** [reqnroll/Reqnroll#16] [Ideas]
Currently when using for instance `Table.CreateInstance<SomeClass>()` internally all ValueRetrievers are stored in a static collection. This means that the first call will fix the ValueRetrievers for the entire test run.

Given an application with a data store and a scenario such as the following:

```gherkin
Given the following customers exists
| Name |
| ABC  |
| DEF  |
When a request comes in to delete the customer ABC 
Then the following customers exist
| Name |
| ABC  |
```

When the record stored actually contains the data ABC then implementation is relatively straightforward. But imagine that the Customer record is actually just some long difficult string (for instance `easdjkf87ewrkhjasdh789ewrkjsasf` with some other non-relevant data. Writing those in scenario's is not something I would want.
So a translation from the identifier used in the scenario and the actual scenario is needed. We use a ValueRetriever (and a StepArgumentTransformation) that is given a per scenario container to generate a mapping between in this case `ABC` and `easdjkf87ewrkhjasdh789ewrkjsasf`, specifically:
```csharp
[Binding]
public class MyHooks
{
    private MappingContext _context;

    public MyHooks(MappingContext context)
    {
        _context = context;
    }

    [BeforeScenario]
    public void SetupTestUsers()
    {
        Service.Instance.ValueRetrievers.Register(new CustomerNameRetriever(_context));
    }
}
```

The above is just a simplification of what we actually do, btw, but I hope the point comes across.

The above will **work perfectly fine** when all of the scenarios are executed in **sequence** as the Register call will overwrite the previous registration.
It breaks down when scenarios are executed in **parallel** as the ValueRetriever used most likely is given a context belonging to a different scenario, whereas the StepArgumentTransformation used will be given the proper context.

------

The issue is that the TEHelpers internally use the Reqnroll.Assist.Service.Instance method, which creates a single static instace that stores a list of ValueComparers and ValueRetrievers. A bit like how the ScenarioContext and FeatureContext used to be static as well.

I'd like to change this to something like a TableHelpers class that is injected through the DI. Which would mean that instead of writing table.CreateSet<>(), you'd write tablehelpers.CreateSet<>(table). The TableHelpers would then have an instance of the Service injected, which would make the ValueRetrievers changeable per scenario, **I think**.

Is this an approach that is viable?

(see also https://github.com/SpecFlowOSS/SpecFlow/issues/2492)

**Comment by gasparnagy** [comment]
This is a good initiative and I was also thinking about that already.

Without going into the details, basically the challenge is that tables are detached from the scenario execution context, but for performing the conversions we would need to have a scenario execution context. This is the reason why we cannot use `[StepArgumentTransformation]`s as well. 

I see two approaches:

1. We make tables attached to the scenario execution context (in most of the cases the Table/DataTable instance is anyway created by Rqnroll, so it could immediately attach it to the context) and make retrievers/comparers scenario context aware (not a static method).
2. We move the `CreateSet` / `CompareToSet` / etc. methods to a service class (maybe to `ScenarioContext`), so they would be called like `_scenarioContext.CreateSet<SomeClass>(myDataTable)`.

Both options have pros and cons... I don't know. 

What do you think?

**Reply by ajeckmans** [reply]
I think option 1 would open the way for people to start manipulating the table in really weird ways, so I'm not sure that is a road we would want to go down.
Option 2 would work only if that `CreateSet` would also use an instance of `Service` that is stored in the `ScenarioContext`.

My initial thought was to inject `Service` into the DI and update all code that touches `Service.Instance` internally to either be a method on the `Service` class (or rather some wrapper around it) or to resolve the `Service` instance from DI as well.

I tried to move all of the extension methods to the `Service` class, but that will break all of the existing static extension method calls and any workaround I tried which would require major updates to some code bases I've worked on in the past 5 years. And I think it is best to not force this on existing code bases (which I guess was the rationale for keeping the `ScenarioContext.Current`, but making it obsolete, as well?).

**Reply by gasparnagy** [reply]
Makes sense.

Maybe we could keep the existing CreateSet, etc. extension methods and let them use the static registration (and the `Service.Instance`) for backwards compatibility and build a new set of methods on an injectable service class. Probably for the sake of clarity I would create a brand new class for the latter. This would serve as a "front-end" API class, we can still share the code logic in internal infrastructure classes.

Also if we do this step, we should consider renaming the frontend API methods. `CompareToSet` is not really a compare method in a sense that it returns true or false, but rather an assertion method, so calling it `AssertXXX` or `ShouldXXX` would be maybe a better choice.

Do you have the energy to make a very draft prototype, just with a single method and other shortcuts if necessary, to test the concept? Of course I'm here to help if you get stuck.

**Reply by ajeckmans** [reply]
It has become annoying enough for us that the company I work for at the moment is willing to spend effort in fixing this. But probably not within the week as I need to free some time to get this rolling.

**Reply by gasparnagy** [reply]
If the idea #18 would be also implemented, the usage of the outlined solution would be even easier.

**Reply by ajeckmans** [reply]
I feel like both scenario's would need to be addressed. Even if you implement property injection, you'd still need to support ctor injection as that is what some will use throughout the code base.

**Reply by gasparnagy** [reply]
Yes, definitely. The property injection is just an optional possibility.

**Reply by ajeckmans** [reply]
I failed to get the test suite working on my laptop, so I had to get a hold of another laptop., but I managed to get started working on a prototype.
 Time during working hours will not be available until April, so it will be spare time only and thus a bit more slow going.

**Reply by gasparnagy** [reply]
No worries @ajeckmans. Thanks for your efforts. (Making the dev setup easier for contributors is one of my other priority goal BTW.)

**Reply by ajeckmans** [reply]
@gasparnagy see #108
I still cannot get all tests to run, but it is way better than it was two months ago!

**Reply by gasparnagy** [reply]
@ajeckmans I would like to learn about the ones that are not running, but let's wait for #109 to be merged first, because I hope that will further improve the situation.

**Comment by orifn1** [comment]
it looks like I have a similar issue with **ValueRetrievers** but another approach
https://github.com/reqnroll/Reqnroll/issues/155

**Reply by ajeckmans** [reply]
yes, though this uses documented features, although ones that are not thread-safe and as such enforce that the tests cannot run in parallel anymore.

I've decided to explore an alternative approach to the one in this PR. That would enable us to fix having to create both StepArgumentTransformations and ValueRetrievers, which both have similar, but not entirely overlapping goals. If it will take too long I'm going to finish this PR though.

**Comment by orifn1** [comment]
Hello, 
any updates?

**Reply by gasparnagy** [reply]
Not yet. But it is on the list. I think what we need is a "big redesign" of what we really want. I can help in that, but during the summer my availability is more limited, so maybe this will be more active again in September.
